skorch/callbacks/logging.py

```diff
@@ -116,97 +117,171 @@ class NeptuneLogger(Callback):
 
     Parameters
     ----------
-    experiment : neptune.experiments.Experiment
-      Instantiated ``Experiment`` class.
+    run : neptune.new.Run
+      Instantiated ``Run`` class.
 
     log_on_batch_end : bool (default=False)
       Whether to log loss and other metrics on batch level.
 
     close_after_train : bool (default=True)
-      Whether to close the ``Experiment`` object once training
+      Whether to close the ``Run`` object once training
       finishes. Set this parameter to False if you want to continue
-      logging to the same Experiment or if you use it as a context
+      logging to the same run or if you use it as a context
       manager.
 
     keys_ignored : str or list of str (default=None)
       Key or list of keys that should not be logged to Neptune. Note that in
       addition to the keys provided by the user, keys such as those starting
       with ``'event_'`` or ending on ``'_best'`` are ignored by default.
 
+    base_namespace: str
+      Namespace (folder) under which all metadata logged by the ``NeptuneLogger``
+      will be stored. Defaults to "training".
+
     Attributes
     ----------
-    first_batch_ : bool
-        Helper attribute that is set to True at initialization and changes
-        to False on first batch end. Can be used when we want to log things
-        exactly once.
-
     .. _Neptune: https://www.neptune.ai
 
     """
 
     def __init__(
             self,
-            experiment,
+            run,
             log_on_batch_end=False,
             close_after_train=True,
             keys_ignored=None,
+            base_namespace='training',
     ):
-        self.experiment = experiment
+        self.run = run
         self.log_on_batch_end = log_on_batch_end
         self.close_after_train = close_after_train
         self.keys_ignored = keys_ignored
 
-    def initialize(self):
-        self.first_batch_ = True
+        if base_namespace.endswith("/"):
+            self._base_namespace = base_namespace[:-1]
+        else:
+            self._base_namespace = base_namespace
 
+    @property
+    def _metric_logger(self):
+        return self.run[self._base_namespace]
+
+    @staticmethod
+    def _get_obj_name(obj):
+        return type(obj).__name__
+
+    def initialize(self):
         keys_ignored = self.keys_ignored
         if isinstance(keys_ignored, str):
             keys_ignored = [keys_ignored]
         self.keys_ignored_ = set(keys_ignored or [])
         self.keys_ignored_.add('batches')
         return self
 
+    def on_train_begin(self, net, X, y, **kwargs):
+
+        self._metric_logger['model/model_type'] = NeptuneLogger._get_obj_name(net.module_)
+        self._metric_logger['model/summary'] = NeptuneLogger._model_summary_file(net.module_)
+
+        self._metric_logger['config/optimizer'] = NeptuneLogger._get_obj_name(net.optimizer_)
+        self._metric_logger['config/criterion'] = NeptuneLogger._get_obj_name(net.criterion_)
```

Comment:
Just a minor note: In skorch, nets can have more than one module, criterion, and optimizer. Here is an examplary snippet that shows how to retrieve all of modules and criteria:

https://github.com/skorch-dev/skorch/blob/5ec195c8eb20a85d8e657595593f44248732b5b9/skorch/net.py#L848-L850

The same logic can be applied for optimizers.

It's fine to keep it as is, though, since I assume only few users make use of it and from all skorch classes, only the GPyTorch classes in `probabilistic.py` add custom modules (the likelihood).

Reply:
@BenjaminBossan could you point me to a code example using multiple modules? What in such case would be logged by `net.module_`, etc?

Reply:
As mentioned, this is an edge case, so feel free to ignore it, maybe leave a `# TODO` comment.

In case you want to tackle it, this test probably best demonstrates what it means if a user sets a custom module:

https://github.com/skorch-dev/skorch/blob/5ec195c8eb20a85d8e657595593f44248732b5b9/skorch/tests/test_net.py#L2662-L2671

As you can see, they can give any name like `mymodule_` to the attribute but we can find it inside of `self._modules`. The `module_` is still set but it just means that there is more than one `nn.Module`. There is no telling which one is the "main" module, so it would be best to log all of them, not sure if there is a limit in Neptune.